### PR TITLE
added IntronicSpliceSite, SpliceDonor, SpliceAcceptor effects

### PR DESCRIPTION
**Pull request by iskandr**

### CODE REVIEW

**Comment by iskandr** [comment]
Fixes https://github.com/hammerlab/varcode/issues/16 and partially fixes https://github.com/hammerlab/varcode/issues/1 (still need to support exonic splice variants)

File: varcode/core_logic.py [hunk 1]
```diff
@@ -381,6 +383,78 @@ def infer_exonic_effect(variant, transcript):
...
+            assert exon.start <= end and exon.end >= start, \
+                "Expected interval %d:%d to overlap with exon %s" % (
+                    start, end, exon)
+            return 0, exon
+
+        if best_distance > distance:
+            best_distance = distance
+            best_exon = exon
+
+    return best_exon, best_exon
```
Comment (timodonnell) [inline varcode/core_logic.py]: best_distance, best_exon ?

File: varcode/core_logic.py [hunk 2]
```diff
@@ -381,6 +383,78 @@ def infer_exonic_effect(variant, transcript):
...
+    best_exon = None
+    for exon in exons:
+        if exon.start > end:
+            # distance from end of variant to start of exon
+            # (when variant starts before the exon)
+            distance = exon.start - end
+        elif exon.end < start:
+            distance = start - exon.end
+        else:
+            assert exon.start <= end and exon.end >= start, \
```
Comment (timodonnell) [inline varcode/core_logic.py]: Given the above conditionals, this assert can't fail, but I guess that's what asserts are for

File: varcode/core_logic.py [hunk 3]
```diff
@@ -381,6 +383,78 @@ def infer_exonic_effect(variant, transcript):
...
         transcript)
 
+def find_nearest_exon(start, end, exons):
+    """
+    Finds nearest exon to an intronic variant, returns the
+    distance to that exon, along with the Exon object itself.
+    """
+    best_distance = float("inf")
+    best_exon = None
+    for exon in exons:
```
Comment (timodonnell) [inline varcode/core_logic.py]: i'd consider putting this logic in a distance_between_intervals(start1, end1, start2, end2) function

File: varcode/variant_collection.py [hunk 4]
```diff
@@ -125,29 +125,33 @@ def variant_effects(self, raise_on_error=True):
...
             for variant
             in self.records
         ]
 
-    def effects_to_string(self):
+    def effects_to_string(self, *args, **kwargs):
         """
         Create a long string with all transcript effects for each mutation,
         grouped by gene (if a mutation affects multiple genes).
         """
```
Comment (timodonnell) [inline varcode/variant_collection.py]: a note in the docstring about where the args and kwargs are sent would be helpful

File: varcode/variant_collection.py [hunk 5]
```diff
@@ -163,8 +167,8 @@ def effects_to_string(self):
                 lines.append("  Highest Priority Effect: %s" % best)
         return "\n".join(lines)
 
-    def print_effects(self):
+    def print_effects(self, *args, **kwargs):
         """
         Print all variants and their transcript effects (grouped by gene).
         """
```
Comment (timodonnell) [inline varcode/variant_collection.py]: similarly re args and kwargs

File: varcode/core_logic.py [hunk 6]
```diff
@@ -381,6 +383,78 @@ def infer_exonic_effect(variant, transcript):
...
+            return 0, exon
+
+        if best_distance > distance:
+            best_distance = distance
+            best_exon = exon
+
+    return best_exon, best_exon
+
+
+def infer_intronic_effect(variant, transcript, nearest_exon, distance_to_exon):
```
Comment (timodonnell) [inline varcode/core_logic.py]: nit: I'd consider not passing around the distance_to_exon value. it's cheap to compute given the variant and the exon, and passing it around opens an opportunity for inconsistency (would want to have a helper function to do this though)

**Comment by timodonnell** [comment]
module comments LGTM

File: varcode/core_logic.py (hunk 3 above)
Comment (iskandr) [inline varcode/core_logic.py]: :+1:

File: varcode/core_logic.py (hunk 2 above)
Comment (iskandr) [inline varcode/core_logic.py]: This is more of a documenting assert, to remind the reader what the else branch means.

File: varcode/core_logic.py (hunk 1 above)
Comment (iskandr) [inline varcode/core_logic.py]: Good catch, thanks. I think this calls for a unit test or two.

File: varcode/core_logic.py (hunk 6 above)
Comment (iskandr) [inline varcode/core_logic.py]: Added a helper method to all Locus objects (Gene, Transcript, Exon) to make this easy to compute.

**Comment by iskandr** [comment]
I moved `find_nearest_exon` into PyEnsembl, see: https://github.com/hammerlab/pyensembl/pull/39/

File: varcode/variant_collection.py (hunk 4 above)
Comment (iskandr) [inline varcode/variant_collection.py]: :+1:

File: varcode/variant_collection.py (hunk 5 above)
Comment (iskandr) [inline varcode/variant_collection.py]: :+1:

**Pull request merged**